hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -3166,24 +3166,33 @@ private TableGroup consumeCrossJoin(SqmCrossJoin<?> sqmJoin, TableGroup lhsTable
 	}
 
 	private TableGroup consumeEntityJoin(SqmEntityJoin<?> sqmJoin, TableGroup lhsTableGroup, boolean transitive) {
+		final List<Predicate> predicates = new ArrayList<>();
 		final EntityPersister entityDescriptor = resolveEntityPersister( sqmJoin.getReferencedPathSource() );
 
 		final SqlAstJoinType correspondingSqlJoinType = sqmJoin.getSqmJoinType().getCorrespondingSqlJoinType();
+		final boolean canUseInnerJoins = correspondingSqlJoinType == SqlAstJoinType.INNER || correspondingSqlJoinType == SqlAstJoinType.CROSS;
 		final TableGroup tableGroup = entityDescriptor.createRootTableGroup(
-				correspondingSqlJoinType == SqlAstJoinType.INNER || correspondingSqlJoinType == SqlAstJoinType.CROSS,
+				canUseInnerJoins,
 				sqmJoin.getNavigablePath(),
 				sqmJoin.getExplicitAlias(),
 				null,
-				() -> (predicate) -> additionalRestrictions = combinePredicates( additionalRestrictions, predicate ),
-				this
-		);
+				() -> (predicate) -> {
+					if ( canUseInnerJoins || correspondingSqlJoinType == SqlAstJoinType.RIGHT ) {
+						additionalRestrictions = combinePredicates( additionalRestrictions, predicate );
+					}
+					else {
+						predicates.add( predicate );
```

Comment:
@beikov you said this should be done for everything but cross joins, but i rather suspect that it should only be done for left joins (not sure about the full join)